TCP related and other fixes

rveerama1 opened the pull request:
This PR contains fixes related to TCP issues and other network related issues.

rveerama1 committed:
net: context: Remove duplicate call

k_delayed_work_cancel(&context->tcp->fin_timer) called twice
immediately one after other.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: context: Handle failure case of tcp backlog entry

In case of failed to get source address from the net packet,
release the assgined tcp backlog entry. Otherwise it will
never be freed.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: tcp: Do not handle packet re-transmission in TCP ACK

Zephyr doesn't have luxury to create re-transmit timer per packet. So
it has different methods to handle packets in queue. But re-sending
packets on valid ack messages causing issues.

E.g. A TCP node sent two packets (packet-1, packet-2). Peer replied
two ACKs (ACK-1 and ACK-2), and these two ACK's are at rx_thread
queue. Now ACK-1 is handled and reference of packet-1 is freed
from sent list. Then if condiftion (valid ACK and connection
state is ESTABLISHED) notices that, sent list is not empty.
Restart the timer, modify sent flag and resend packets in a
list. Here packet-2 is sent again, even though ACK-2 is already
received. Situation is worse if there are more packets in the
list.

So only start the re-transmit timer in-case queue is not empty. It
allows rx_thread to handle all incoming packets (in this e.g ACKs).
When the re-trasmit timer expires, it sends the packets which
are left in queue.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: tcp: Provide local address in TCP reset message preparation

If context is bound to IPv6 unspecified addresss and some port
number, then unspecified address is passed in TCP reset packet
message preparation. Eventually packet dropped at the peer.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

pfalcon commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	  to choose default interface. Otherwise first interface will be the
+	  default interface.
+
+config NET_DEFAULT_IF_NONE
```
The name of the option contradicts its description: the option says that there's not default interface, while description says the first is default.

pfalcon reviewed (changes requested):
At least one issue (with NET_DEFAULT_IF_NONE option naming) requires change IMHO.

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -367,11 +367,29 @@ struct net_if *net_if_lookup_by_dev(struct device *dev)
+
 	if (__net_if_start == __net_if_end) {
-		return NULL;
+		return iface;
```
I wouldn't make this change. The original code makes (more) clear that the return value is NULL.

jukkar replied:
I agree with Paul

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1246,6 +1246,14 @@ NET_CONN_CB(tcp_established)
 		return NET_DROP;
 	}
 
+	/* Increment the ack */
```
This is *not* a correct change. First, application gets a chance to process incoming data. An application may reject or fail to process data (e.g., due to lack of memory). ACK gets sent only after that, and includes outcomes of the processing status (e.g., updated/not updated receive window, etc.).

If you look thru the history, you'll likely find that the opposite change was done, which you now try to revert.

pfalcon commented on subsys/net/ip/tcp.c:
```diff
@@ -961,32 +961,14 @@ bool net_tcp_ack_received(struct net_context *ctx, u32_t ack)
 		valid_ack = true;
 	}
 
-	/* No need to re-send stuff we are closing down */
```
Comment for the commit message:

> Now ACK's are in rx_thread queue or ACK for first packet received.

I'm not sure I understand this phrase. Can you please clarify the description in commit message?

pfalcon replied:
Actually, whole this part is not very clear:

> Now ACK's are
in rx_thread queue or ACK for first packet received. Then
net_tcp_ack_received() function will delete the first packet.
Because it's valid ACK and connection state is in ESTABLISHED state
sent list is not empty. It restarts the timer and resend packets
in queue. In e.g. case Zephyr sends second packet in a list
immediately.

pfalcon commented on subsys/net/ip/tcp.c:
```diff
@@ -661,12 +661,47 @@ int net_tcp_prepare_ack(struct net_tcp *tcp, const struct sockaddr *remote,
 	return -EINVAL;
 }
 
+static void get_src_addr_ptr(struct net_tcp *tcp,
```
I don't think this is a right change, see https://github.com/zephyrproject-rtos/zephyr/issues/5481

rveerama1 replied:
I tried to find helper function for this purpose. May be I missed it. I will take a look.

pfalcon commented on subsys/net/lib/app/net_app.c:
```diff
@@ -562,27 +562,45 @@ struct net_context *select_client_ctx(struct net_app_ctx *ctx,
 
 #if defined(CONFIG_NET_APP_SERVER)
 #if defined(CONFIG_NET_TCP)
-static struct net_context *get_server_ctx(struct net_app_ctx *ctx,
```
Commit message:

~~~
E.g. One net_app_ctx and two net_contexts. 'For' loop goes through
     array of net_contexts and searching for matching context.
     Let's say caller required net_context is second in the
~~~ 

What's the purpose of that intent? Please reformat.

pfalcon replied:
Also, that description also doesn't sound too clear to me (but I'm not familiar with net_app_ctx internals, maybe that's it).

pfalcon commented:
My usual comment applies: this PR packs the full spectrum of changes - from obviously correct to obviously incorrect, with anything inbetween.

At this point, I'm keen to say that Github reviews are failing for us, and we should go back to Gerrit which enforces "one commit per review" rule. <--- @nashif: sharing an opinion.

nashif commented:
> At this point, I'm keen to say that Github reviews are failing for us, and we should go back to Gerrit which enforces "one commit per review" rule. <--- @nashif: sharing an opinion.

I am not sure how this can be true, you could have 1000s of commits in 1 single gerrit change set, same thing as a PR in GH. 

Review per commit, and not as a PR, just the same way you would have done in gerrit. 

i agree we should not put too much unrelated stuff into one PR, it can be confusing, commits in 1 single PR need to be related.

And finally, No, we are not going back to gerrit.

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	  default interface.
+
+config NET_DEFAULT_IF_NONE
+	bool "Default interface is first interface"
```
This could be shorter so what about "First available interface"

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	  default interface.
+
+config NET_DEFAULT_IF_ETHERNET
+	bool "Ethernet interface is the default interface"
```
Just "Ethernet" here

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	depends on NET_L2_ETHERNET
+
+config NET_DEFAULT_IF_BLUETOOTH
+	bool "Bluetooth interface is the default interface"
```
"Bluetooth"

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	depends on NET_L2_BT
+
+config NET_DEFAULT_IF_IEEE802154
+	bool "IEEE802.15.4 interface is the default interface"
```
"IEEE 802.15.4"

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	depends on NET_L2_IEEE802154
+
+config NET_DEFAULT_IF_OFFLOAD
+	bool "OFFLOAD interface is the default interface"
```
"Offloaded interface"

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -349,6 +349,42 @@ config NET_BUF_USER_DATA_SIZE
+	depends on NET_L2_OFFLOAD
+
+config NET_DEFAULT_IF_DUMMY
+	bool "Dummy interface is the default interface"
```
"Dummy testing interface"

jukkar commented:
Most of the patches seem not to be related to each other but I am not sure if it is worth sending them one by one. But if needed at least TCP stuff could be a separate PR, and interface selection and net-app fixes. So this could be split into three PRs.

rveerama1 commented on subsys/net/ip/net_context.c:
```diff
@@ -1253,14 +1263,6 @@ NET_CONN_CB(tcp_established)
-	}
-
-	send_ack(context, &conn->remote_addr, false);
-
```
@pfalcon "ret = packet_received(conn, pkt, context->tcp->recv_user_data);" Sending ack doesn't depend on returned value. Irrespective of return value, ACK will be prepared and send.

pfalcon replied:
> Irrespective of return value, ACK will be prepared and sent.

Yes, an ACK packet will be sent in either case. But *contents* of the ACK packet are expected to depend on app's receive callback (i.e. receive callback may affect data put into an ACK reply), specifically, the receive window: 19ff9636930f1ceabae3461a43c6c1c78cf30253 .

> I noticed an issue related to this while trying with RPL BR sample. ...

Thanks for the clear description of the problem you saw. I would say, this description should go into the commit message.

But then, we found yet another issue with callback-based data processing. And it can be classified as a case of https://cwe.mitre.org/data/definitions/405.html - for a small request, "http_prepare_and_send() create new packets until all the payload filled in buffers", and all that happens in a not very appropriate place - synchronous callback from the network stack, which affects the stack performance.

I'm afraid, we'd need to find another solution, and such solution exists - BSD Sockets. How it works is that there's decoupling between packet reception of packets by network stack and consumption by an application - they both run in separate threads and communicate via queue. There's also the flow control implemented, mandated by TCP, using the concept of "receive window": when a packet is just received from peer peer and put into the queue, the recv_wnd is decreased (and communicated back in an ACK packet), signaling the peer that it has decreased quota to send more data (which eventually may get to 0). recv_wnd is increased back only when data is actually consumed from the queue.

Due to the fact that we have usecases of both updating and not update recv_wnd, the (optional) update happens in recv callback. Thus, this patch would break this processing, and will break cases of communication large amounts of data (megabytes) over sockets.

My suggestion is to separate this patch into a separate PR, and further analyze and test it from different sides, but opinion is that we'd need to find an alternative way to address this issue. Thanks.

rveerama1 replied:
> But then, we found yet another issue with callback-based data processing. And it can be classified as a case of https://cwe.mitre.org/data/definitions/405.html - for a small request, "http_prepare_and_send() create new packets until all the payload filled in buffers", and all that happens in a not very appropriate place - synchronous callback from the network stack, which affects the stack performance.

I am experimenting that in my WIP PR of RPL BR. Running a separate thread to reply websocket messages. Checking for proper solution.
https://github.com/zephyrproject-rtos/zephyr/pull/5035/commits/b61d0a5d5f3d2dd3c954cf579828725b2391860e#diff-780b2211d7789425a2584af4afc52b77R1881

rveerama1 replied:
Overall, let's take it away patch from this PR, time being I will keep it in https://github.com/zephyrproject-rtos/zephyr/pull/5035. I will create a separate PR for this and list issues over there.

rveerama1 reviewed (commented):
Added some comments, please have a look.

rveerama1 commented:
Split the patches into separate PRs (https://github.com/zephyrproject-rtos/zephyr/pull/6073, https://github.com/zephyrproject-rtos/zephyr/pull/6074).
@pfalcon : addressed your comments, except duplicate function to get src socket ptr. if issue https://github.com/zephyrproject-rtos/zephyr/issues/5481 provides any generic API, I will re-use that.

pfalcon commented:
> @pfalcon : addressed your comments, except duplicate function to get src socket ptr. if issue #5481 provides any generic API, I will re-use that.

Thanks. Please let me check the current situation with that, and come up with a specific suggestion.

rveerama1 commented:
Dropped the ACK sending before payload handler patch.

jukkar reviewed (approved):
LGTM

jukkar commented:
@pfalcon could you re-review the PR as now the controversial "ACK sending before payload handler" commit is dropped from this set.

pfalcon commented on subsys/net/ip/tcp.c:
```diff
@@ -661,12 +661,47 @@ int net_tcp_prepare_ack(struct net_tcp *tcp, const struct sockaddr *remote,
+#endif
+
+end:
+	addr = &tcp->context->local;
```
This is a dead statement, and apparently doesn't do what was intended by the author (what?).

pfalcon commented:
> @pfalcon could you re-review the PR as now the controversial "ACK sending before payload handler" commit is dropped from this set.

Sure, as my last comment 2 comments above yours says, the last concern is still get_src_addr_ptr() function added in https://github.com/zephyrproject-rtos/zephyr/pull/6058/commits/3146804e6c1f8197f36dd4ed5d993d25ec4a6978#diff-a278935f8d5d5e0b4424ca3e1182d48cR664

@jukkar, @rveerama1 : And I'm glad that I took another look at it. Here're concerns with it:

1. It contains an obviously dead statement, assignment to a local var without further use (like logical thinko re: logic that assignment should perform).
2. As I mentioned, my concern (based on the name of the function) was that it would be 4th function with almost similar purpose in the codebase (variability of src/dst addr_ptr/no-ptr), tracked in #5481. But looking in more detail into it, turns out "get_src_addr_ptr" isn't what it does! What it appears to do is to *copy* `struct sockaddr` into `struct sockaddr_ptr`, with an adhoc exception for a case when `struct sockaddr` is NULL (coded improperly per p.1). In particular, it doesn't copy "src" addr, it can copy *any* addr. Summing: the problem with that function is the naming: it's named based on the usage *in the single place*, instead of  based on what it *does*. Yet, it's named generically enough to cause confusion re: if it can be reused or should reuse existing similar functions. Given that it's called from a single place, I'd suggest to just put the code directly there. Failing that, I'd suggest to make a generic `copy_sockaddr_to_sockaddr_ptr()` function, and still remove adhoc check `if (!local) ... &tcp->context->local` check from it. @jukkar , maybe you can suggest if such a generic copy function already exists.

rveerama1 commented:
@pfalcon updated PR. Let me know if I address your comments.

pfalcon reviewed (approved):
> updated PR. Let me know if I address your comments.

Looks good, thanks much for reworking it!
